### ISSUE TITLE
PSRule fails to expand resources presumably if resource uses map() function in output

### ISSUE THREAD

**Issue by AlexanderSehr** [microsoft/PSRule#2591]
**Description of the issue**

In AVM, I recently commited an update to a module (private endpoint) that returns an array of resource IDs in its output using the following function
```bicep
output networkInterfaceResourceIds string[] = map(privateEndpoint.properties.networkInterfaces, nic => nic.id)
```

Following this, I started implementing this version in 2 modules (Key Vault & Storage Account), that not only uses this module, but also passes this output through to the user. For [example](https://github.com/Azure/bicep-registry-modules/blob/5999204d51145f2fe6b10003feacd5068ab195bb/avm/res/key-vault/vault/main.bicep#L395-L403):

```bicep
@description('The private endpoints of the key vault.')
output privateEndpoints privateEndpointOutputType[] = [
  for (item, index) in (privateEndpoints ?? []): {
    (...)
    networkInterfaceResourceIds: keyVault_privateEndpoints[index].outputs.networkInterfaceResourceIds
  }
]
``` 

Now, in AVM, where we run PSRule tests both on files that deploy private endpoints and not, an error surfaced for the one that does deploy private endpoints, and in turn, uses the output ([Key Vault test example](https://github.com/AlexanderSehr/bicep-registry-modules/actions/runs/11533006094/job/32105429915#step:4:222), [Storage Account test example](https://github.com/Azure/bicep-registry-modules/actions/runs/11533050395/job/32105524578#step:4:232)). The error looks as follows:

_`Failed to expand bicep source '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep' was not valid. An error occurred evaluating expression '[map(reference('privateEndpoint').networkInterfaces, lambda('nic', lambdaVariables('nic').id))]' line 3535. The arguments for 'Map' are not in the expected format or type."`_

I did do some investigation to ensure the output is not just broken, but it looks exactly like I'd exact it to:
```json
[
    {
        "name": "pep-<short>ssawaf001-blob-0",
        "resourceId": "/subscriptions/<subscriptionId>/resourceGroups/dep-<short>-storage.storageaccounts-ssawaf-rg/providers/Microsoft.Network/privateEndpoints/pep-<short>ssawaf001-blob-0",
        "groupId": "blob",
        "customDnsConfigs": [
            {
                "fqdn": "<short>ssawaf001.blob.core.windows.net",
                "ipAddresses": [
                    "10.0.0.4"
                ]
            }
        ],
        "networkInterfaceResourceIds": [
            "/subscriptions/<subscriptionId>/resourceGroups/dep-<short>-storage.storageaccounts-ssawaf-rg/providers/Microsoft.Network/networkInterfaces/pep-<short>ssawaf001-blob-0.nic.0c73db37-6089-4ba7-9d31-86827a6ae489"
        ]
    }
]
```

and no deployment test, no matter if a private endpoint deployment nor without fails.

So, right now I'm not sure what to make of this, as the `map()` function in the output seems to work as intended, yet PSRule seems to stumble over it in this context. Any help would be greatly appreciated. 

**To Reproduce**

Steps to reproduce the issue:

- Run the PSRule Reliability tests on either one of those to templates:
  - [Key Vault - WAF test](https://github.com/AlexanderSehr/bicep-registry-modules/blob/users/alsehr/Gosselin-l_keyvault_key_moreOutputs/avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep)
  - [Storage Account - WAF test](https://github.com/Azure/bicep-registry-modules/blob/users/alsehr/saUpdates20241026/avm/res/storage/storage-account/tests/e2e/waf-aligned/main.test.bicep)

> Note: We use a PSRule configuration file that e.g.  excludes resources with `dep` in the name from validation. 
> - [General config](https://github.com/Azure/bicep-registry-modules/blob/users/alsehr/saUpdates20241026/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/dep-suppress.Rule.yaml)
> - ['dep' config](https://github.com/Azure/bicep-registry-modules/blob/users/alsehr/saUpdates20241026/avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml)

**Expected behaviour**

Rather simple - I'd expect PSRule to successfully process the corresponding test file without an error.

**Error output**

Key Vault WAF module test example
_`Failed to expand bicep source '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep' was not valid. An error occurred evaluating expression '[map(reference('privateEndpoint').networkInterfaces, lambda('nic', lambdaVariables('nic').id))]' line 3535. The arguments for 'Map' are not in the expected format or type."`_

Storage Accont WAF module test example
_`Error: Failed to expand bicep source '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/storage/storage-account/tests/e2e/waf-aligned/main.test.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/storage/storage-account/tests/e2e/waf-aligned/main.test.bicep' was not valid. An error occurred evaluating expression '[map(reference('privateEndpoint').networkInterfaces, lambda('nic', lambdaVariables('nic').id))]' line 2890. The arguments for 'Map' are not in the expected format or type."`_

**Module in use and version:**

- Module: PSRule
- Version: **2.9.0**

Captured output from `$PSVersionTable`:

```text
  Name                           Value
  ----                           -----
  PSVersion                      7.4.5
  PSEdition                      Core
  GitCommitId                    7.4.5
  OS                             Ubuntu 22.04.5 LTS
  Platform                       Unix
  PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
  PSRemotingProtocolVersion      2.3
  SerializationVersion           1.1.0.1
  WSManStackVersion              3.0
```

**Additional context**

**Comment by BernieWhite** [comment]
Thanks for reporting the issue @AlexanderSehr.

**Comment by BernieWhite** [comment]
@AlexanderSehr Should be fixed in PSRule for Azure v1.40.0-B0029.

**Comment by AlexanderSehr** [comment]
Hey @BernieWhite,
that's great news. And quick too.
Looking forward to the release 💪 
Thank you sir.
